### PR TITLE
micro transactions support added

### PR DESCRIPTION
**Pull request by taburet**

### CODE REVIEW

File: core/src/main/java/com/orientechnologies/orient/core/db/document/ODatabaseDocumentAbstract.java [hunk 1]
```diff
@@ -1478,10 +1486,18 @@ public boolean isPrefetchRecords() {
 
       // SEARCH IN LOCAL TX
       ORecord record = getTransaction().getRecord(rid);
-      if (record == OTransactionRealAbstract.DELETED_RECORD)
+      if (record == OBasicTransaction.DELETED_RECORD)
```
Comment (taburet) [inline core/src/main/java/com/orientechnologies/orient/core/db/document/ODatabaseDocumentAbstract.java]: assertion added

File: core/src/main/java/com/orientechnologies/orient/core/db/document/ODatabaseDocumentAbstract.java [hunk 2]
```diff
@@ -1478,10 +1486,18 @@ public boolean isPrefetchRecords() {
...
         return null;
 
+      if (record == null) {
+        if (microTransaction != null && microTransaction.isActive()) {
+          record = microTransaction.getRecord(rid);
+          if (record == OBasicTransaction.DELETED_RECORD)
+            return null;
+        }
+      }
+
```
Comment (taburet) [inline core/src/main/java/com/orientechnologies/orient/core/db/document/ODatabaseDocumentAbstract.java]: method removed completely

File: core/src/main/java/com/orientechnologies/orient/core/index/OClassIndexManager.java [hunk 3]
```diff
@@ -300,85 +322,6 @@ private boolean processSingleIndexDelete(final OIndex<?> index, final Set<String
...
   }
 
-  private ODocument checkIndexedPropertiesOnCreation(final ODocument record, final Collection<OIndex<?>> indexes) {
-    ODocument replaced = null;
-
-    final TreeMap<OIndex<?>, List<Object>> indexKeysMap = new TreeMap<OIndex<?>, List<Object>>();
-
-    for (final OIndex<?> index : indexes) {
-      if (index.getInternal() instanceof OIndexUnique) {
-        OIndexRecorder indexRecorder = new OIndexRecorder((OIndexUnique) index.getInternal());
```
Comment (taburet) [inline core/src/main/java/com/orientechnologies/orient/core/index/OClassIndexManager.java]: OIndexRecorder deleted

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/OMicroTransaction.java [hunk 4]
```diff
@@ -0,0 +1,705 @@
...
+
+  private final int id;
+
+  private final Map<ORID, ORecordOperation>                       recordOperations      = new LinkedHashMap<>();
+  private final Map<String, OTransactionIndexChanges>             indexOperations       = new LinkedHashMap<>();
+  private final Map<ORID, List<OTransactionRecordIndexOperation>> recordIndexOperations = new HashMap<>();
+
+  private final Map<ORID, ORID> updatedRids      = new HashMap<>();
+  private final Set<ODocument>  changedDocuments = new HashSet<>();
+
```
Comment (taburet) [inline core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/OMicroTransaction.java]: It was a bug, changed documents tracking is fixed. Good you noticed it.

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/OAbstractPaginatedStorage.java [hunk 5]
```diff
@@ -2100,6 +2236,18 @@ private void makeRollback(OTransaction clientTx, Exception e) {
 
   }
 
+  private void makeRollback(OMicroTransaction microTransaction, Exception e) {
+    // WE NEED TO CALL ROLLBACK HERE, IN THE LOCK
+    OLogManager.instance().debug(this, "Error during micro-transaction commit, micro-transaction will be rolled back (tx-id=%d)", e,
+        microTransaction.getId());
+    rollback(microTransaction);
+    if (e instanceof OException)
+      throw ((OException) e);
```
Comment (taburet) [inline core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/OAbstractPaginatedStorage.java]: fixed

File: core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/OMicroTransaction.java [hunk 6]
```diff
@@ -0,0 +1,705 @@
...
+  private final ODatabaseDocumentInternal database;
+  private final OAbstractPaginatedStorage storage;
+
+  private final int id;
+
+  private final Map<ORID, ORecordOperation>                       recordOperations      = new LinkedHashMap<>();
+  private final Map<String, OTransactionIndexChanges>             indexOperations       = new LinkedHashMap<>();
+  private final Map<ORID, List<OTransactionRecordIndexOperation>> recordIndexOperations = new HashMap<>();
+
```
Comment (taburet) [inline core/src/main/java/com/orientechnologies/orient/core/storage/impl/local/OMicroTransaction.java]: javadoc added

File: core/src/main/java/com/orientechnologies/orient/core/index/OIndexManagerShared.java [hunk 7]
```diff
@@ -521,7 +511,7 @@ public void run() {
         }
 
         try {
-          recreateIndexes(storage, indexesToRebuild, newDb);
+          recreateIndexes(indexesToRebuild, newDb);
         } finally {
           if (storage instanceof OAbstractPaginatedStorage) {
```
Comment (laa) [inline core/src/main/java/com/orientechnologies/orient/core/index/OIndexManagerShared.java]: Switch to unsafe mode and force sync is not needed anymore, let's check if this method is not used anywhere I propose to delete it.

File: core/src/main/java/com/orientechnologies/orient/core/index/OIndexManagerShared.java [hunk 8]
```diff
@@ -556,44 +545,62 @@ private void recreateIndexes(OStorage storage, Collection<ODocument> idxs, OData
...
+              indexMetadata.getName());
+          addIndexAsIs(indexDocument, index);
+        } else {
+          OLogManager.instance()
+              .info(this, "Index '%s' is a non-durable automatic index and must be rebuilt", indexMetadata.getName());
+          rebuildNonDurableAutomaticIndex(indexDocument, index, indexMetadata, indexDefinition);
+        }
       } else {
-        addNonAutomaticIndex(idx, index, indexMetadata);
+        if (durable) {
```
Comment (laa) [inline core/src/main/java/com/orientechnologies/orient/core/index/OIndexManagerShared.java]: Why we re-add already existing indexes if they already in index manager, please either provide clear comment in code or delete this code

File: core/src/main/java/com/orientechnologies/orient/core/index/OIndexManagerShared.java (hunk 7 above)
Comment (taburet) [inline core/src/main/java/com/orientechnologies/orient/core/index/OIndexManagerShared.java]: Seems like `OIndexAbstract#loadFromConfiguration` may rebuild indexes on its own. Not sure is it ok to turn the WAL on while it is doing the reindexing. WDYT?

File: core/src/main/java/com/orientechnologies/orient/core/index/OIndexManagerShared.java [hunk 9]
```diff
@@ -361,29 +351,29 @@ public boolean autoRecreateIndexesAfterCrash() {
...
     acquireExclusiveLock();
     try {
-      final Map<String, OIndex<?>> oldIndexes = new HashMap<String, OIndex<?>>(indexes);
+      final Map<String, OIndex<?>> oldIndexes = new HashMap<>(indexes);
 
       clearMetadata();
-      final Collection<ODocument> idxs = document.field(CONFIG_INDEXES);
-      final Locale locale = getServerLocale();
+      final Collection<ODocument> indexDocuments = document.field(CONFIG_INDEXES);
+      getServerLocale();
```
Comment (taburet) [inline core/src/main/java/com/orientechnologies/orient/core/index/OIndexManagerShared.java]: fixed, calls removed

File: core/src/main/java/com/orientechnologies/orient/core/index/OIndexManagerShared.java (hunk 8 above)
Comment (taburet) [inline core/src/main/java/com/orientechnologies/orient/core/index/OIndexManagerShared.java]: comment added

File: core/src/main/java/com/orientechnologies/orient/core/index/OIndexManagerShared.java (hunk 7 above)
Comment (taburet) [inline core/src/main/java/com/orientechnologies/orient/core/index/OIndexManagerShared.java]: Unsafe mode removed. Index rebuilding left as it is with minor modifications, will create a separate issue.

**Pull request merged**